pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -499,19 +499,30 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 			changes.CreateSandbox = false
 			return changes
 		}
+
+		// Get the containers to start, excluding the ones that succeeded if RestartPolicy is OnFailure.
+		var containersToStart []int
+		for idx, c := range pod.Spec.Containers {
+			if pod.Spec.RestartPolicy == v1.RestartPolicyOnFailure && containerSucceeded(&c, podStatus) {
+				continue
+			}
+			containersToStart = append(containersToStart, idx)
+		}
+		// We should not create a sandbox for a Pod if initialization is done and there is no container to start.
+		if len(containersToStart) == 0 {
+			_, _, done := findNextInitContainerToRun(pod, podStatus)
+			if done {
+				changes.CreateSandbox = false
+				return changes
```

Comment:
perhaps this can be simplified. If you will remove the return statement from here and the new condition with the call to `findNextInitContainerToRun`, everything suppose to be working just fine. If there are Init containers, `changes` will be returned from the next if condition with the `changes.CreateSandbox = false` and `changes.NextInitContainerToStart` populated. It will save on additional call to `findNextInitContainerToRun ` and will minimize the new changes

Reply:
In this branch `if createPodSandbox`, it's supposed to (re-)create sandbox because it doesn't exist or it's dead or a changed is detected, and we are only adding exceptions to stop that if all things are done, like the existing exception: `if !shouldRestartOnFailure(pod) && attempt != 0 && len(podStatus.ContainerStatuses) != 0`  for the case that a Pod whose RestartPolicy is Never and have created sandbox and containers before.
It's supposed to either return `CreateSandbox = false` or `CreateSandbox = true` with `NextInitContainerToStart` or `ContainersToStart`.
If the returned changes is `CreateSandbox = false` with `NextInitContainerToStart`, there would be no sandbox for the init container.